### PR TITLE
WIP TypeScript refactor

### PR DESCRIPTION
**Pull request by kincaidoneil**
### Big Things

- Refactor in TypeScript a la xrp-asym-server
- Add balance logic; options for `maximum`, `settleTo` and `settleThreshold`
- Add channel watcher to claim channels if the sender tries to settle, if it's profitable
- Support for unidirectional sending: server will accumulate funds for an account up to the cost of an on-chain transaction, and only when it hits that threshold, open an outgoing payment channel (the client could either open a paychan to the server to front their cost, or keep fulfilling and wait until enough funds to it accumulated) [(More details on our outline/proposal for this here)](https://github.com/interledger/rfcs/issues/443).

### Little Things

- Fee estimation logic for different operations
- Add maxPacketAmount limiting
- Add support for `ilp-logger`
- Use store wrapper from xrp-asym-server
- Simplify account schema for now; use change detection to trigger updates to store
- Simplify BTP subprotocols to `info`: retrieves the Ethereum address the peer wants to be paid at; and `machinomy`: sends serialized Machinomy payments (also used to prove channel ownership)
- Prevent multiple simultaneous funding transactions
- Added upstream Machinomy fixes for payment validation and best claim
- Add account blocking
- Add max packet amount config
- Other miscellaneous fixes

### What's Next

- This has not been tested at all, so we expect a lot of bugs (we wanted to push this out as a work in progress first to solicit feedback). We're currently in the process of writing unit tests and pushing out fixes, but we'd love any feedback or help with tests!
- We're also looking to add black-box/integration tests as well so the whole stack have better security guarantees (https://github.com/machinomy/machinomy/issues/232)
- Open question for us: what is the optimal DoS prevention scheme? With the xrp plugins, the client has to escrow some multiple of what the server locks up (10x right now). In the case of Ethereum, we're not sure that's a sufficient mitigation due to higher fees, because if the server gets a payment for less than the cost to claim the channel, but still forwards the payment, the attacker can force the server to lose money (even though the attacker would likely lose more). Should a client with an incoming paychan have to prefund (to cover the server's fee if they needed to claim the channel) before the server begins crediting them any balance? Looking for ideas on the best approach

(Credit to @karzak for a ton of help with this!)

### CODE REVIEW

**Comment by emschwartz** [comment]
For the future, please submit more smaller PRs. It's nearly impossible to do a proper review of a 50+ commit PR

File: src/__tests__/stream.test.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  const INCOMING_FEE = convert('0.00045', Unit.Eth, Unit.Gwei)
+  const RECEIVER_MAX_BALANCE = 0
+
+  let actualReceived = new BigNumber(0)
+
+  let serverConn: IlpStream.Connection
+  let serverStream: IlpStream.DataAndMoneyStream
+
+  let clientConn: IlpStream.Connection
+  let clientStream: IlpStream.DataAndMoneyStream
```
Comment (emschwartz) [inline src/__tests__/stream.test.ts]: Why are these declared above where they're used?

File: src/account.ts [hunk 2]
```diff
@@ -0,0 +1,843 @@
...
+  /**
+   * How settlement works:
+   * - Determine the maximum amount to settle, or settlement "budget", and immediately add it to the balance
+   * - Each settlement operation will spend down from this amount, subtracting sent claims and fees
+   * - `sendClaim` and `fundOutgoingChannel` take the settlement budget, and return the amount leftover
+   *    - e.g. if we couldn't open a channel because the fee > amountToSettle, it will error and just
+   *      return the original amount
+   *    - e.g. if it sent a claim for the entire amountToSettle, it will return 0 because there's
+   *      nothing leftover
+   * - This is really cool, because the amount to settle can "flow" from one settlement method to another,
```
Comment (emschwartz) [inline src/account.ts:267]: What are the different settlement methods you're referring to here?

File: src/account.ts [hunk 3]
```diff
@@ -0,0 +1,843 @@
...
+      try {
+        // 1) Try to send a claim: spend the channel down entirely before funding it
+        amountLeftover = await this.sendClaim(settlementBudget)
+        // 2) Open or deposit to a channel if it's necessary to send anything leftover
+        if (amountLeftover.lte(0)) throw amountLeftover
+        amountLeftover = await this.fundOutgoingChannel(amountLeftover)
+        // 3) Try to send a claim again since we may have more funds in the channel
+        if (amountLeftover.lte(0)) throw amountLeftover
+        amountLeftover = await this.sendClaim(amountLeftover)
+      } catch (leftover) {
```
Comment (emschwartz) [inline src/account.ts]: It's strange to throw things that aren't errors. What if `this.sendClaim` actually throws an error? Seems like amountLeftover would be wrong

File: src/account.ts [hunk 4]
```diff
@@ -0,0 +1,843 @@
...
+      return settlementBudget
+    } catch (err) {
+      this.master._log.error(`Failed to send claim to ${this.account.accountName}: ${err.message}`)
+      throw settlementBudget
+    }
+  }
+
+  async handleData (message: BtpPacket, dataHandler?: DataHandler): Promise<BtpSubProtocol[]> {
+    const info = getSubprotocol(message, 'info')
+    const ilp = getSubprotocol(message, 'ilp')
```
Comment (emschwartz) [inline src/account.ts:589]: nit: don't need to check for both if it's only valid to have included one of them (and one will be ignored if you set both anyway)

File: src/account.ts [hunk 5]
```diff
@@ -0,0 +1,843 @@
...
+          const oldClaim = this.account.bestIncomingClaim
+
+          // If the account already has a channel, client must be using that channel
+          // Calculating profitability across multiple incoming channels sounds like a receipe for disaster
+          if (oldClaim && oldClaim.channelId !== claim.channelId) {
+            throw new Error(`account ${this.account.accountName} must use channel ${oldClaim.channelId}`)
+          }
+
+          // Always fetch updated channel state from network
+          const channel = await fetchChannel(this.master._web3, claim.channelId)
```
Comment (emschwartz) [inline src/account.ts]: Does it really make sense to make that API call every time you get a claim in? If the channels are unidirectional you shouldn't need to do that unless you think they might have put more money in the channel and you want to check if they did

File: src/account.ts [hunk 6]
```diff
@@ -0,0 +1,843 @@
...
+          const contractAddress = (await getNetwork(this.master._web3)).unidirectional.address
+          // @ts-ignore http://web3js.readthedocs.io/en/1.0/web3-utils.html#soliditysha3
+          const paymentDigest = Web3.utils.soliditySha3(contractAddress, claim.channelId, claim.value)
+          const senderAddress = this.master._web3.eth.accounts.recover(paymentDigest, claim.signature)
+          const isValidSignature = senderAddress === channel.sender
+          if (!isValidSignature) {
+            throw new Error(`signature is invalid, or sender is using contracts on a different network`)
+          }
+
+          if (new BigNumber(claim.value).lte(0)) {
```
Comment (emschwartz) [inline src/account.ts]: nit: check this before checking the signature

File: src/account.ts [hunk 7]
```diff
@@ -0,0 +1,843 @@
...
+            claimIncrement = claimIncrement.minus(this.master._incomingChannelFee)
+            if (claimIncrement.lt(0)) {
+              throw new Error(`claim value of ${format(newClaimValue, Unit.Wei)} is insufficient to cover the initiation fee of ` +
+                `${format(this.master._incomingChannelFee, Unit.Wei)}`)
+            }
+
+            this.master._store.set(channelKey, this.account.accountName)
+            this.master._log.trace(`Incoming channel ${claim.channelId} is now linked to account ${this.account.accountName}`)
+
+            // TODO web3.js beta 36 event bugs: https://github.com/ethereum/web3.js/issues/1916
```
Comment (emschwartz) [inline src/account.ts:774]: This seems kind of important, no?

File: src/utils/queue.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import 'source-map-support/register'
```
Comment (emschwartz) [inline src/utils/queue.ts]: Apparently there are arguments against including this in libraries, cause it doesn't play nicely with webpacking things

File: src/__tests__/stream.test.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  })
+
+  await t.notThrowsAsync(serverStream!.receiveTotal(AMOUNT_TO_SEND), 'client streamed the total amount of packets to the server')
+
+  // Wait 10 seconds for sender to finish sending the remaining payment channel claims
+  await new Promise(resolve => {
+    setTimeout(resolve, 10000)
+  })
+
+  t.true(actualReceived.gte(AMOUNT_TO_SEND.minus(INCOMING_FEE)), 'server received at least as much money as the client sent')
```
Comment (emschwartz) [inline src/__tests__/stream.test.ts:144]: Would this fail if the network fees increase?

File: src/account.ts (hunk 7 above)
Comment (kincaidoneil) [reply src/account.ts:774]: The channel watcher polls at regular intervals, so it should still work fine. The event subscriptions are just nice because it'd trigger it to claim the channel instantly (if it was profitable). I'll add it back in after the Web3 bug is fixed.

File: src/__tests__/stream.test.ts (hunk 9 above)
Comment (kincaidoneil) [reply src/__tests__/stream.test.ts:144]: That test should still work. `INCOMING_FEE` is a fixed initiation fee that the connector charges when it gets a claim from a new channel (to cover the cost of them claiming it, although right now it's not dynamic to the cost of the claim tx).

However, the tests _would_ fail if the fee to open a channel was greater than the `settleTo` amount in the sender's balance configuration (since even on the client, the fee to open a channel is added to/debited from the balance). That said, the prefund amount is `0.002` eth, so the transaction fees should almost always be less than that.

File: src/account.ts (hunk 4 above)
Comment (kincaidoneil) [reply src/account.ts:589]: Yeah, it'd be weird for the packet to include both. I think having both checks is useful though in case they sent some random subprotocol name that we don't support.

File: src/account.ts [hunk 10]
```diff
@@ -0,0 +1,843 @@
...
+        } else {
+          this.master._log.info(`Successfully deposited ${format(value, Unit.Wei)} to channel ${channelId} for account ${this.account.accountName}`)
+        }
+      } else {
+        this.master._log.trace(`No on-chain funding transaction required to settle ${format(settlementBudget, Unit.Wei)} with account ${this.account.accountName}`)
+      }
+
+      return settlementBudget
+    } catch (err) {
+      this.master._log.error(`Failed to fund outgoing channel to ${this.account.accountName}: ${err.message}`)
```
Comment (sentientwaffle) [inline src/account.ts:497]: Since the `settlementBudget` is sometimes thrown, this error doesn't make sense (and `settlementBudget.message` is probably `undefined`). As evan mentioned, throwing amounts is hard to follow. If there are different amounts/states that need to be returned, maybe return an object with one of 2+ fields.

File: src/plugin.ts [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  _handleMoney (from: string, message: BtpPacket): Promise<BtpSubProtocol[]> {
+    return this._getAccount(from).handleMoney(message, this._moneyHandler)
+  }
+
+  _handlePrepareResponse = async (
+    destination: string,
+    responsePacket: IlpPacket.IlpPacket,
+    preparePacket: IlpPacket.IlpPacket
+  ): Promise<void> =>
+    this._getAccount(destination).handlePrepareResponse(preparePacket, responsePacket)
```
Comment (sentientwaffle) [inline src/plugin.ts:140]: `handlePrepareResponse` doesn't return a `Promise` -- this only type checks because `_handlePrepareResponse` is marked `async`. Also, why use `_handlePrepareResponse = async (...)` instead of `async _handlePrepareResponse (...`.

File: src/account.ts (hunk 5 above)
Comment (kincaidoneil) [reply src/account.ts]: ~~Maybe not, but it might be a premature optimization. I think it's safer to do it each claim. For example, if the channel was claimed outside of the plugin and the state was not fetched, new claims would still be credited.~~

File: src/__tests__/watcher.test.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+  await web3.eth.sendTransaction(tx)
+
+  await new Promise(resolve => {
+    setInterval(async () => {
+      const wasClaimed = await contract.methods.isAbsent(channelId).call()
+
+      if (wasClaimed) {
+        t.pass()
+        resolve()
```
Comment (sentientwaffle) [inline src/__tests__/watcher.test.ts:78]: `clearInterval()` so that this doesn't keep running after the promise is resolved.

**Review by sentientwaffle — CHANGES_REQUESTED** [review]

File: src/account.ts (hunk 10 above)
Comment (kincaidoneil) [reply src/account.ts:497]: I updated it so it only rejects the promise in those cases rather than throwing. Is there a cleaner way to do this?

File: src/plugin.ts (hunk 11 above)
Comment (kincaidoneil) [reply src/plugin.ts:140]: Fixed the promise ✅

When I define it that way, TypeScript throws an error that it's an instance member property and not an instance member function (I think it has to do with this: https://github.com/interledgerjs/ilp-plugin-mini-accounts/blob/master/src/index.ts#L65).

File: src/account.ts (hunk 10 above)
Comment (sentientwaffle) [reply src/account.ts:497]: Hm, I see. That resolves the error message issue, but I'm still concerned about rejecting with non-error objects. It's a valid but unexpected pattern, which leads to subtle errors. For example:

```javascript
      } catch (leftover) {
        // If no money was sent, rejections should return the original amount
        if (BigNumber.isBigNumber(leftover)) {
          amountLeftover = leftover
        }
      }
```

If the caught error isn't a `BigNumber`, it is discarded -- is that the correct behavior?

For clarity, I'd suggest having the function return `{amountLeftover, success: true|false}` and never reject with an amount. Instead, the caller would check `success` and behave accordingly. It's a little more verbose, but I think it's easier to follow.

File: src/account.ts [hunk 13]
```diff
@@ -413,18 +418,16 @@ export default class EthereumAccount {
...
       } else {
         this.master._log.trace(`No on-chain funding transaction required to settle ${format(settlementBudget, Unit.Wei)} with account ${this.account.accountName}`)
       }
-
-      return settlementBudget
     } catch (err) {
       this.master._log.error(`Failed to fund outgoing channel to ${this.account.accountName}: ${err.message}`)
-      throw settlementBudget
+    } finally {
+      return settlementBudget
```
Comment (sentientwaffle) [inline src/account.ts:499]: nit: I think it would be clearer if this `return` was in the `catch` with the log line, since if no error is thrown it will return from within the `try` block.

File: src/account.ts (hunk 5 above)
Comment (kincaidoneil) [reply src/account.ts]: Disregard. You're right. In my testing, processing a claim took 300-500ms even with a 4ms latency connection, which--given the incoming claims are queued--is unacceptable if settlement is triggered on every packet. Furthermore, even if no funding tx is required, the channel state is also queried before signing an outgoing claim.

I'll add caching of the channel state for both incoming and outgoing channels. This assumes that a user doesn't deposit to or claim their channel outside of the plugin, but that's probably a reasonable assumption.

File: src/account.ts (hunk 13 above)
Comment (kincaidoneil) [reply src/account.ts:499]: Are you sure? Maybe I'm misinterpreting, but I think there are a number of cases where it wouldn't return the settlement budget from within the try (unless I added a return within both the try and catch).

File: src/account.ts (hunk 2 above)
Comment (kincaidoneil) [reply src/account.ts:267]: `sendClaim` and `fundOutgoingChannel`

**Comment by kincaidoneil** [comment]
@sentientwaffle @emschwartz Is this good to merge? Still intend to only release as betas, but I think it's stable enough for master.

**Review by emschwartz — APPROVED** [review]

**Review by sentientwaffle — APPROVED** [review]
